### ISSUE TITLE
On NUMA machines, half of the shards are moderately slower, due to .text being on the other node

### ISSUE THREAD

**Issue by michoecho** [scylladb/scylla#9294]
Related to #9252. While sharing .text across NUMA nodes is nowhere near as bad as inter-node cache thrashing, it's definitely not performance-neutral. For example, in a write-to-memory throughput test on Ampere A1, moving .text (and nothing else) to the other node reduces the IPC from ~1 to ~0.65, and reduces the throughput accordingly.
Unlike with thrashing, the strength of this effect does not seem to scale with the number of cores: it's always 30-40%, no matter if we are using 2 cores or 64 cores. (And, unlike thrashing, I think it doesn't affect the cores on the same node as .text at all, but I can't confirm that due to #9292 interfering).

If we care about NUMA performance, it would be good to fix it.

This problem would be solved by giving the shards on the second node a separate physical copy of .text, and I wonder if it's something that can be hacked in.
What if we compiled Scylla as a relocatable object, and at startup duplicated the mappings of all sections except .text, but mapped in a second physical copy of .text? Then shards would jump into the second copy of .text at startup. I think this would break debugging (AFAIK symbols are ELF-relative, not segment-relative), but could work otherwise (AFAIK unwind tables are segment-relative). Or is it too hacky to even consider?

**Comment by avikivity** [comment]
There's lots of prior art about it, we should look at existing work. This is something best solved in the kernel.

**Comment by avikivity** [comment]
Although, I don't see how Linux can map the same virtual address to two physical addresses, since all threads share the same page table. It could be done in principle (by unsharing the page tables) but in practice it would be a monumental effort.

**Comment by avikivity** [comment]
> What if we compiled Scylla as a relocatable object, and at startup duplicated the mappings of all sections except .text, but mapped in a second physical copy of .text? Then shards would jump into the second copy of .text at startup. I think this would break debugging (AFAIK symbols are ELF-relative, not segment-relative), but could work otherwise (AFAIK unwind tables are segment-relative). Or is it too hacky to even consider?

Function pointers break it.

**Comment by avikivity** [comment]
Some prior art: https://lwn.net/Articles/45334/

**Comment by avikivity** [comment]
What's L3 size on that system? Because code should be cached in L3.

Perhaps we should relieve L3 by storing only on copy of read-only metadata (such as token_metadata) per NUMA node, not per shard. /cc @bhalevy

**Comment by michoecho** [comment]
> Function pointers break it.

Right. But maybe only partially? Executing half of the code from remote .text should be better than executing all code from remote .text.

> What's L3 size on that system? Because code should be cached in L3.

This system has 64KiB of L1i, 64KiB of L1d and 1MiB of L2 per core and 32MiB of L3 (called System Level Cache in the docs) shared between cores. But I'm not sure if this L3 can cache memory belonging to the other node. I think I was doing some synthetic experiments which suggested that it didn't.
(I was checking if cores can share (to other cores on the same node) cache lines belonging to the other node to other cores (they can't), but I think I also concluded that L3 doesn't cache them. I might be misremembering though.)

**Comment by michoecho** [comment]
Or, maybe more accurately, it *is* in L3, but in remote L3, so we still need a remote access.

**Comment by avikivity** [comment]
> > Function pointers break it.
> 
> Right. But maybe only partially? Executing half of the code from remote .text should be better than executing all code from remote .text.
> 
> > What's L3 size on that system? Because code should be cached in L3.
> 
> This system has 64KiB of L1i, 64KiB of L1d and 1MiB of L2 per core and 32MiB of L3 (called System Level Cache in the docs) shared between cores. But I'm not sure if this L3 can cache memory belonging to the other node. I think I was doing some synthetic experiments which suggested that it didn't.


I think it's indeed a memory-side cache (so after coherency) and doesn't cache remote data. If so I don't think we can support such a machine well. Other NUMAs have real L3s.

> (I was checking if cores can share (to other cores on the same node) cache lines belonging to the other node to other cores (they can't),

They should be able to share if those lines are read-only.

> but I think I also concluded that L3 doesn't cache them. I might be misremembering though.)

**Comment by avikivity** [comment]
According to https://amperecomputing.com/wp-content/uploads/2021/06/Altra_Rev_A1_DS_v1.10_20210612-1.pdf (section 2.5) this is a memory-side cache, and so won't cache remote data.

**Comment by avikivity** [comment]
Well, we can improve things by relieving L2 instead of L3 by reducing the amount of duplicated read-only metadata. But without a real L3, we can't recommend NUMA configurations of this processor.

**Comment by michoecho** [comment]
> They should be able to share if those lines are read-only.

So here's a experiment for that: we allocate a 8KiB array on the remote node, and a 1MiB array on the local node. We spawn one thread on core 0 to read the remote array in a loop (so the entire array is in its L1 cache), and we spawn a thread on core 1, which reads the 1MiB array and then the 8KiB array in a loop. This is done so that the 1MiB array forces the 8KiB array out of core 1's L2 cache.

`perf stat` shows that core 1 does remote accesses for the remote array, so apparently core 0 isn't sharing its cache with core 1, even though it's read-only.  

```
#include <cstdlib>
#include <cstring>
#include <numaif.h>
#include <numa.h>
#include <iostream>
#include <memory>
#include <thread>
#include <vector>

uint64_t *remote;

size_t REMOTE_SIZE = 8*1024;
size_t LOCAL_SIZE = 1024*1024;

int main(int argc, const char** argv) {
        remote = reinterpret_cast<uint64_t*>(numa_alloc_onnode(REMOTE_SIZE, 1));
        for (size_t i = 0; i < REMOTE_SIZE/8; i+=8) {
                asm volatile("str %[dst], %[src]\n" : [src]"=m"(remote[i]) : [dst]"r"(i));
        }

        std::vector<std::thread> threads(2);
        threads[1] = std::thread([] {
                while (true) {
                        for (size_t i = 0; i < REMOTE_SIZE/8; i+=8) {
                                uint64_t x;
                                asm volatile("ldr %[dst], %[src]\n" : [dst]"=r"(x) : [src]"m"(remote[i]));
                        }
                }
        });
        threads[0] = std::thread([] {
                auto buf = std::unique_ptr<uint64_t[]>(reinterpret_cast<uint64_t*>(numa_alloc_onnode(LOCAL_SIZE, 0)));
                for (size_t i = 0; i < LOCAL_SIZE/8; i+=8) {
                        uint64_t x;
                        asm volatile("str %[dst], %[src]\n" : [src]"=m"(buf[i]) : [dst]"r"(i));
                }
                while (true) {
                        for (size_t i = 0; i < LOCAL_SIZE/8; i+=8) {
                                uint64_t x;
                                asm volatile("ldr %[dst], %[src]\n" : [dst]"=r"(x) : [src]"m"(buf[i]));
                        }
                        for (size_t i = 0; i < REMOTE_SIZE/8; i+=8) {
                                uint64_t x;
                                asm volatile("ldr %[dst], %[src]\n" : [dst]"=r"(x) : [src]"m"(remote[i]));
                        }
                }
        });

        cpu_set_t cpuset;
        CPU_ZERO(&cpuset);
        CPU_SET(0, &cpuset);
        int rc = pthread_setaffinity_np(threads[1].native_handle(), sizeof(cpu_set_t), &cpuset);
        if (rc != 0) { std::cerr << "Error calling pthread_setaffinity_np: " << rc << "\n"; }

        CPU_ZERO(&cpuset);
        CPU_SET(1, &cpuset);
        rc = pthread_setaffinity_np(threads[0].native_handle(), sizeof(cpu_set_t), &cpuset);
        if (rc != 0) { std::cerr << "Error calling pthread_setaffinity_np: " << rc << "\n"; }

        for (auto& t : threads) {
                t.join();
        }
        return 0;
}
```

**Comment by avikivity** [comment]
> > They should be able to share if those lines are read-only.
> 
> So here's a experiment for that: we allocate a 8KiB array on the remote node, and a 1MiB array on the local node. 

I misread your comment. I thought you asked about a local-node line among local cores.


Although I think the problem we see is more subtle. If a line is available in L2 of any core in the system (local or remote), we should read from L2, not memory. I think the remote access is for the directory, since the directory for a line is local to the node that owns the line.

I think this happens:

c1n0: read line
c1n0: miss L1/L2, ask directory on n1
n1: directory says c1n0, fetch line from c1n0
c1n0: there's the line
n1: mark c0n0 as another sharer
n1: return line to c0n0
c0n0: thanks
c0n0: I nearly fell asleep waiting

so it's two round trips. Or maybe n1 tells c1n0 to send the line to c0n0, and it's just one round trip.

With a real L3, the read is intercepted by L3 and the directory is not consulted.

**Comment by dimaqq** [comment]
A user's perspective:
1. Kernel support would be ideal; other multithreaded programs would benefit too
2. Until then, running a process per node may more practical
   * I'm not sure if Linux will share the .text segment for same executable on NUMA... maybe there's a trick around that.
   * It's possible to share synchronisation primitives even between unrelated process (not parent-child) [1]

[1] https://stackoverflow.com/questions/35265781/share-futex-between-unrelated-processes
